File: web/share/js/kvm/hid.js
/*****************************************************************************
#                                                                            #
#    KVMD - The main PiKVM daemon.                                           #
#                                                                            #
#    Copyright (C) 2018-2024  Maxim Devaev <mdevaev@gmail.com>               #
#                                                                            #
#    This program is free software: you can redistribute it and/or modify    #
#    it under the terms of the GNU General Public License as published by    #
#    the Free Software Foundation, either version 3 of the License, or       #
#    (at your option) any later version.                                     #
#                                                                            #
#    This program is distributed in the hope that it will be useful,         #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of          #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
#    GNU General Public License for more details.                            #
#                                                                            #
#    You should have received a copy of the GNU General Public License       #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.  #
#                                                                            #
*****************************************************************************/


"use strict";


import {tools, $, $$$} from "../tools.js";
import {wm} from "../wm.js";

import {Keyboard} from "./keyboard.js";
import {Mouse} from "./mouse.js";


export function Hid(__getGeometry, __recorder) {
	var self = this;

	/************************************************************************/

	var __state = null;
	var __keyboard = null;
	var __mouse = null;

	var __init__ = function() {
		__keyboard = new Keyboard(__recorder.recordWsEvent);
		__mouse = new Mouse(__getGeometry, __recorder.recordWsEvent);

		let hidden_attr = null;
		let visibility_change_attr = null;

		if (typeof document.hidden !== "undefined") {
			hidden_attr = "hidden";
			visibility_change_attr = "visibilitychange";
		} else if (typeof document.webkitHidden !== "undefined") {
			hidden_attr = "webkitHidden";
			visibility_change_attr = "webkitvisibilitychange";
		} else if (typeof document.mozHidden !== "undefined") {
			hidden_attr = "mozHidden";
			visibility_change_attr = "mozvisibilitychange";
		}

		if (visibility_change_attr) {
			document.addEventListener(
				visibility_change_attr,
				function() {
					if (document[hidden_attr]) {
						__releaseAll();
					}
				},
				false
			);
		}

		window.addEventListener("pagehide", __releaseAll);
		window.addEventListener("blur", __releaseAll);

		tools.el.setOnClick($("hid-connect-switch"), __clickConnectSwitch);
		tools.el.setOnClick($("hid-reset-button"), __clickResetButton);

		for (let el_shortcut of $$$("[data-shortcut]")) {
			tools.el.setOnClick(el_shortcut, function() {
				let ask = false;
				let confirm_id = el_shortcut.getAttribute("data-shortcut-confirm");
				if (confirm_id) {
					ask = $(confirm_id).checked;
				}
				let codes = el_shortcut.getAttribute("data-shortcut").split(" ");
				if (ask) {
					wm.confirm("Do you want to press this hotkey?", codes.join(" + ")).then(function(ok) {
						if (ok) {
							__emitShortcut(codes);
						}
					});
				} else {
					__emitShortcut(codes);
				}
			});
		}

		tools.storage.bindSimpleSwitch($("hid-sysrq-ask-switch"), "hid.sysrq.ask", true);

		tools.el.setOnClick($("hid-jiggler-switch"), __clickJigglerSwitch);
	};

	/************************************************************************/

	self.setSocket = function(ws) {
		if (!ws) {
			self.setState(null);
		}
		__keyboard.setSocket(ws);
		__mouse.setSocket(ws);
	};

	self.setState = function(state) {
		if (state) {
			if (!__state) {
				__state = {"keyboard": {}, "mouse": {}};
			}
			if (state.enabled !== undefined) {
				__state.enabled = state.enabled; // Currently unused, always true
			}
			if (__state.enabled !== undefined) {
				for (let key of ["online", "busy", "connected", "jiggler"]) {
					if (state[key] !== undefined) {
						__state[key] = state[key];
					}
				}
				for (let hid of ["keyboard", "mouse"]) {
					if (state[hid] === undefined) {
						state[hid] = {}; // Add some stubs for processing
					}
					for (let key of ["online", "outputs", (hid === "keyboard" ? "leds" : "absolute")]) {
						__state[hid][key] = state[hid][key];
					}
				}
				if (state.connected !== undefined) {
					tools.feature.setEnabled($("hid-connect"), (__state.connected !== null));
					$("hid-connect-switch").checked = !!__state.connected;
				}
				if (state.jiggler !== undefined) {
					tools.feature.setEnabled($("hid-jiggler"), __state.jiggler.enabled);
					$("hid-jiggler-switch").checked = __state.jiggler.active;
				}
				if (state.keyboard.outputs !== undefined) {
					__updateKeyboardOutputs(__state.keyboard.outputs);
				}
				if (state.mouse.outputs !== undefined) {
					__updateMouseOutputs(__state.mouse.outputs, __state.mouse.absolute); // Follows together
				}
				if (
					state.keyboard.online !== undefined || state.keyboard.leds !== undefined
					|| state.online !== undefined || state.busy !== undefined
				) {
					__keyboard.setState(__state.keyboard.online, __state.keyboard.leds, __state.online, __state.busy);
				}
				if (
					state.mouse.online !== undefined || state.mouse.absolute !== undefined
					|| state.online !== undefined || state.busy !== undefined
				) {
					__mouse.setState(__state.mouse.online, __state.mouse.absolute, __state.online, __state.busy);
				}
				if (state.online !== undefined || state.busy !== undefined) {
					tools.radio.setEnabled("hid-outputs-keyboard-radio", (__state.online && !__state.busy));
					tools.radio.setEnabled("hid-outputs-mouse-radio", (__state.online && !__state.busy));
					tools.el.setEnabled($("hid-connect-switch"), (__state.online && !__state.busy));
				}
			}
		} else {
			__state = null;
			tools.radio.setEnabled("hid-outputs-keyboard-radio", false);
			tools.radio.setEnabled("hid-outputs-mouse-radio", false);
			tools.el.setEnabled($("hid-connect-switch"), false);
			tools.el.setEnabled($("hid-mouse-squash-switch"), false);
			tools.el.setEnabled($("hid-mouse-sens-slider"), false);
		}
		tools.el.setEnabled($("hid-reset-button"), __state);
		tools.el.setEnabled($("hid-jiggler-switch"), __state);
	};

	var __updateKeyboardOutputs = function(outputs) {
		let avail = outputs.available;
		if (avail.length > 0) {
			let el = $("hid-outputs-keyboard-box");
			let avail_json = JSON.stringify(avail);
			if (el.__avail_json !== avail_json) {
				let html = "";
				for (let kv of [
					["USB",  "usb"],
					["PS/2", "ps2"],
					["Off",  "disabled"],
				]) {
					if (avail.includes(kv[1])) {
						html += tools.radio.makeItem("hid-outputs-keyboard-radio", kv[0], kv[1]);
					}
				}
				el.innerHTML = html;
				tools.radio.setOnClick("hid-outputs-keyboard-radio", () => __clickOutputsRadio("keyboard"));
				el.__avail_json = avail_json;
			}
			tools.radio.setValue("hid-outputs-keyboard-radio", outputs.active);
		}
		tools.feature.setEnabled($("hid-outputs-keyboard"), (avail.length > 0));
	};

	var __updateMouseOutputs = function(outputs, absolute) {
		let has_relative = null;
		let has_relative_squash = null;
		let avail = outputs.available;
		if (avail.length > 0) {
			let el = $("hid-outputs-mouse-box");
			let avail_json = JSON.stringify(avail);
			if (el.__avail_json !== avail_json) {
				has_relative = false;
				let html = "";
				for (let kv of [
					["Absolute",  "usb",       false],
					["Abs-Win98", "usb_win98", false],
					["Relative",  "usb_rel",   true],
					["PS/2",      "ps2",       true],
					["Off",       "disabled",  false],
				]) {
					if (avail.includes(kv[1])) {
						html += tools.radio.makeItem("hid-outputs-mouse-radio", kv[0], kv[1]);
						has_relative = (has_relative || kv[2]);
					}
				}
				el.innerHTML = html;
				tools.radio.setOnClick("hid-outputs-mouse-radio", () => __clickOutputsRadio("mouse"));
				el.__avail_json = avail_json;
			}
			tools.radio.setValue("hid-outputs-mouse-radio", outputs.active);
			has_relative_squash = (["usb_rel", "ps2"].includes(outputs.active));
		} else {
			has_relative = !absolute;
			has_relative_squash = has_relative;
		}
		if (has_relative !== null) {
			tools.feature.setEnabled($("hid-mouse-squash"), has_relative);
			tools.feature.setEnabled($("hid-mouse-sens"), has_relative);
		}
		tools.feature.setEnabled($("hid-outputs-mouse"), (avail.length > 0));
		tools.el.setEnabled($("hid-mouse-squash-switch"), has_relative_squash);
		tools.el.setEnabled($("hid-mouse-sens-slider"), has_relative_squash);
	};

	var __releaseAll = function() {
		__keyboard.releaseAll();
		__mouse.releaseAll();
	};

	var __emitShortcut = function(codes) {
		return new Promise(function(resolve) {
			tools.debug("HID: emitting keys:", codes);

			let raw_events = [];
			[[codes, true], [codes.slice().reverse(), false]].forEach(function(op) {
				let [op_codes, state] = op;
				for (let code of op_codes) {
					raw_events.push({"code": code, "state": state});
				}
			});

			let index = 0;
			let iterate = () => setTimeout(function() {
				__keyboard.emit(raw_events[index].code, raw_events[index].state);
				++index;
				if (index < raw_events.length) {
					iterate();
				} else {
					resolve(null);
				}
			}, 100);
			iterate();
		});
	};

	var __clickOutputsRadio = function(hid) {
		let output = tools.radio.getValue(`hid-outputs-${hid}-radio`);
		tools.httpPost("api/hid/set_params", {[`${hid}_output`]: output}, function(http) {
			if (http.status !== 200) {
				wm.error("Can't configure HID", http.responseText);
			}
		});
	};

	var __clickJigglerSwitch = function() {
		let enabled = $("hid-jiggler-switch").checked;
		tools.httpPost("api/hid/set_params", {"jiggler": enabled}, function(http) {
			if (http.status !== 200) {
				wm.error(`Can't ${enabled ? "enabled" : "disable"} mouse jiggler`, http.responseText);
			}
		});
	};

	var __clickConnectSwitch = function() {
		let connected = $("hid-connect-switch").checked;
		tools.httpPost("api/hid/set_connected", {"connected": connected}, function(http) {
			if (http.status !== 200) {
				wm.error(`Can't ${connected ? "connect" : "disconnect"} HID`, http.responseText);
			}
		});
	};

	var __clickResetButton = function() {
		wm.confirm("Are you sure you want to reset HID (keyboard & mouse)?").then(function(ok) {
			if (ok) {
				tools.httpPost("api/hid/reset", null, function(http) {
					if (http.status !== 200) {
						wm.error("HID reset error", http.responseText);
					}
				});
			}
		});
	};

	__init__();
}

File: kvmd/apps/localhid/hid.py
# ========================================================================== #
#                                                                            #
#    KVMD - The main PiKVM daemon.                                           #
#                                                                            #
#    Copyright (C) 2018-2024  Maxim Devaev <mdevaev@gmail.com>               #
#                                                                            #
#    This program is free software: you can redistribute it and/or modify    #
#    it under the terms of the GNU General Public License as published by    #
#    the Free Software Foundation, either version 3 of the License, or       #
#    (at your option) any later version.                                     #
#                                                                            #
#    This program is distributed in the hope that it will be useful,         #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of          #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
#    GNU General Public License for more details.                            #
#                                                                            #
#    You should have received a copy of the GNU General Public License       #
#    along with this program.  If not, see <https://www.gnu.org/licenses/>.  #
#                                                                            #
# ========================================================================== #


import asyncio

from typing import Final
from typing import Generator

import evdev
from evdev import ecodes


# =====
class Hid:  # pylint: disable=too-many-instance-attributes
    KEY:          Final[int] = 0
    MOUSE_BUTTON: Final[int] = 1
    MOUSE_REL:    Final[int] = 2
    MOUSE_WHEEL:  Final[int] = 3

    def __init__(self, path: str) -> None:
        self.__device = evdev.InputDevice(path)

        caps = self.__device.capabilities(absinfo=False)

        syns = caps.get(ecodes.EV_SYN, [])
        self.__has_syn = (ecodes.SYN_REPORT in syns)

        leds = caps.get(ecodes.EV_LED, [])
        self.__has_caps = (ecodes.LED_CAPSL in leds)
        self.__has_scroll = (ecodes.LED_SCROLLL in leds)
        self.__has_num = (ecodes.LED_NUML in leds)

        keys = caps.get(ecodes.EV_KEY, [])
        self.__has_keyboard = (
            ecodes.KEY_LEFTCTRL in keys
            or ecodes.KEY_RIGHTCTRL in keys
            or ecodes.KEY_LEFTSHIFT in keys
            or ecodes.KEY_RIGHTSHIFT in keys
        )

        rels = caps.get(ecodes.EV_REL, [])
        self.__has_mouse_rel = (
            ecodes.BTN_LEFT in keys
            and ecodes.REL_X in rels
        )

        self.__grabbed = False

    def is_suitable(self) -> bool:
        return (self.__has_keyboard or self.__has_mouse_rel)

    def set_leds(self, caps: bool, scroll: bool, num: bool) -> None:
        if self.__grabbed:
            if self.__has_caps:
                self.__device.set_led(ecodes.LED_CAPSL, caps)
            if self.__has_scroll:
                self.__device.set_led(ecodes.LED_SCROLLL, scroll)
            if self.__has_num:
                self.__device.set_led(ecodes.LED_NUML, num)

    def set_grabbed(self, grabbed: bool) -> None:
        if self.__grabbed != grabbed:
            getattr(self.__device, ("grab" if grabbed else "ungrab"))()
            self.__grabbed = grabbed

    def close(self) -> None:
        try:
            self.__device.close()
        except Exception:
            pass

    async def poll_to_queue(self, queue: asyncio.Queue[tuple[int, tuple]]) -> None:
        def put(event: int, args: tuple) -> None:
            queue.put_nowait((event, args))

        move_x = move_y = 0
        wheel_x = wheel_y = 0
        async for event in self.__device.async_read_loop():
            if not self.__grabbed:
                # Клавиши перехватываются всегда для обработки хоткеев,
                # всё остальное пропускается для экономии ресурсов.
                if event.type == ecodes.EV_KEY and event.value != 2 and (event.code in ecodes.KEY):
                    put(self.KEY, (event.code, bool(event.value)))
                continue

            if event.type == ecodes.EV_REL:
                match event.code:
                    case ecodes.REL_X:
                        move_x += event.value
                    case ecodes.REL_Y:
                        move_y += event.value
                    case ecodes.REL_HWHEEL:
                        wheel_x += event.value
                    case ecodes.REL_WHEEL:
                        wheel_y += event.value

            if not self.__has_syn or event.type == ecodes.SYN_REPORT:
                if move_x or move_y:
                    for xy in self.__splitted_deltas(move_x, move_y):
                        put(self.MOUSE_REL, xy)
                    move_x = move_y = 0
                if wheel_x or wheel_y:
                    for xy in self.__splitted_deltas(wheel_x, wheel_y):
                        put(self.MOUSE_WHEEL, xy)
                    wheel_x = wheel_y = 0

            elif event.type == ecodes.EV_KEY and event.value != 2:
                if event.code in ecodes.KEY:
                    put(self.KEY, (event.code, bool(event.value)))
                elif event.code in ecodes.BTN:
                    put(self.MOUSE_BUTTON, (event.code, bool(event.value)))

    def __splitted_deltas(self, delta_x: int, delta_y: int) -> Generator[tuple[int, int], None, None]:
        sign_x = (-1 if delta_x < 0 else 1)
        sign_y = (-1 if delta_y < 0 else 1)
        delta_x = abs(delta_x)
        delta_y = abs(delta_y)
        while delta_x > 0 or delta_y > 0:
            dx = sign_x * max(min(delta_x, 127), 0)
            dy = sign_y * max(min(delta_y, 127), 0)
            yield (dx, dy)
            delta_x -= 127
            delta_y -= 127

    def __str__(self) -> str:
        info: list[str] = []
        if self.__has_syn:
            info.append("syn")
        if self.__has_keyboard:
            info.append("keyboard")
        if self.__has_mouse_rel:
            info.append("mouse_rel")
        return f"Hid({self.__device.path!r}, {self.__device.name!r}, {self.__device.phys!r}, {', '.join(info)})"
